### ISSUE TITLE
Satisfaction::satisfy on `AndV` flips the two branches

### ISSUE THREAD

**Issue by afilini** [apoelstra/rust-miniscript#87]
I was recently playing around with Miniscript and I noticed that whenever I try to satisfy a descriptor in the form of `and_v(vc:pk_k(A),c:pk_k(B))`, the two signatures appears to be inverted in the scriptSig/witness.

By inverted I mean that the signature made with key `B` is placed on the stack in such a way that it will be checked against key `A`. I'm honestly not sure if this happens every time an "and_v" is involved or if this is some kind of edge case.

I threw together this quick example to demonstrate the issue. The idea is that I'm creating two public keys and two signatures, then I assign the signature "with many zeros" as if it was made by the public key that also has "many zeros". It sucks but I guess it works to kinda tag the key and signature:

```rust
    let public_keys = vec![
        bitcoin::PublicKey::from_slice(&[2; 33]).expect("key 1"),
        bitcoin::PublicKey::from_slice(&[
            0x02,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
        ]).expect("key 2"),
    ];
    let bitcoin_sigs = vec![(
        // copied at random off the blockchain; this is not actually a valid
        // signature for this transaction; Miniscript does not verify
        secp256k1::Signature::from_str(
            "3045\
             0221\
             00f7c3648c390d87578cd79c8016940aa8e3511c4104cb78daa8fb8e429375efc1\
             0220\
             531d75c136272f127a5dc14acc0722301cbddc222262934151f140da345af177",
        )
        .unwrap(),
        bitcoin::SigHashType::All,
    ), (
        secp256k1::Signature::from_str(
            "3045\
             0221\
             00f700000000000000000000000000000000000000000000000000000000000000\
             0220\
             5300000000000000000000000000000000000000000000000000000000000000",
        )
        .unwrap(),
        bitcoin::SigHashType::All,
    )];

    let descriptor_str = format!(
        "sh(and_v(vc:pk_k({}),c:pk_k({})))",
        public_keys[0], public_keys[1],
    );

    println!("{}", descriptor_str);

    // Descriptor for the output being spent
    let my_descriptor =
        BitcoinDescriptor::from_str(&descriptor_str[..]).expect("parse descriptor string");

    println!("{}", my_descriptor.witness_script());

    let mut sigs = HashMap::<bitcoin::PublicKey, miniscript::BitcoinSig>::new();
    sigs.insert(public_keys[0], bitcoin_sigs[0]);
    sigs.insert(public_keys[1], bitcoin_sigs[1]);

    assert!(my_descriptor.satisfy(&mut tx.input[0], &sigs).is_ok());
    println!("{}", tx.input[0].script_sig);
```

And that's the output:

```
sh(and_v(vc:pk_k(020202020202020202020202020202020202020202020202020202020202020202),c:pk_k(020102030405060708010203040506070800000000000000000000000000000000)))
Script(OP_PUSHBYTES_33 020202020202020202020202020202020202020202020202020202020202020202 OP_CHECKSIGVERIFY OP_PUSHBYTES_33 020102030405060708010203040506070800000000000000000000000000000000 OP_CHECKSIG)
Script(OP_PUSHBYTES_72 3045022100f7c3648c390d87578cd79c8016940aa8e3511c4104cb78daa8fb8e429375efc10220531d75c136272f127a5dc14acc0722301cbddc222262934151f140da345af17701 OP_PUSHBYTES_72 3045022100f7000000000000000000000000000000000000000000000000000000000000000220530000000000000000000000000000000000000000000000000000000000000001 OP_PUSHBYTES_70 21020202020202020202020202020202020202020202020202020202020202020202ad21020102030405060708010203040506070800000000000000000000000000000000ac)
```

So if we ignore the last push that is just the script for the P2SH, we can see that the first signature it is going to check is the "one with many zeros", but it's going to check it against the other public key in the first `OP_CHECKSIGVERIFY`.

The stupid/obvious fix I came up with is just to flip the two branches like this:

```
--- a/src/miniscript/satisfy.rs
+++ b/src/miniscript/satisfy.rs
@@ -580,7 +580,7 @@ impl Satisfaction {
                 let l_sat = Self::satisfy(&l.node, stfr);
                 let r_sat = Self::satisfy(&r.node, stfr);
                 Satisfaction {
-                    stack: Witness::combine(l_sat.stack, r_sat.stack),
+                    stack: Witness::combine(r_sat.stack, l_sat.stack),
                     has_sig: l_sat.has_sig || r_sat.has_sig,
                 }
             }
```

But I guess this could have some side effects that I don't know of.

I would love to work on a proper fix for this, I guess I just need a little bit of help to head to the right direction.

Thanks!

**Comment by apoelstra** [comment]
I think this might be https://github.com/apoelstra/rust-miniscript/pull/85

**Comment by sanket1729** [comment]
Addressed by #85
